### ISSUE TITLE
Version 2.0.0

### ISSUE THREAD

**Issue by nlf** [nlf/dlite#135]
Version 2.0.0 of DLite is ready for testing!

Here's what you can do to help:

First, remove your old installation of DLite

``` sh
dlite stop
sudo dlite uninstall
sudo nfsd stop #unless you're using nfsd outside of DLite
```

You'll also want to edit `/etc/exports` and remove the entry that DLite created

Build from the latest code in the `master` branch (copy the binary to your path if you want. if you installed with homebrew before you'll want to `brew uninstall dlite` first) or download the latest pre-release binary on the [releases page](https://github.com/nlf/dlite/releases/) and install passing the `-v` flag like so:

``` sh
sudo dlite install -v 3.0.0-beta4
```

After the installation completes, run `dlite start` and wait a minute or so. If your internet connection is slow and the version of docker requested in your config is not 1.10.2 it will take longer since on the first boot the docker binary gets downloaded, and it's over 30MB. You'll know it's done and running when `docker ps` works.

Please report any issues you have here and I'll work to get them fixed up before the official release. Thanks!

Edit: You're also welcome to join the [gitter](https://gitter.im/nlf/dlite) for questions or just to say hi

**Comment by ruimarinho** [comment]
I would add:
- `sudo nfsd stop`
- Remove dlite entries from `/etc/exports`.

Some issues:
- Looks like the path to the socket is url encoded.

```
❯ docker ps
Error: request returned Service Unavailable for API route and version http://%2Fvar%2Frun%2Fdocker.sock/v1.21/containers/json, check if the server supports the requested API version
```

This was on my `dlite-err.log`:

```
operation not supported by device
rdmsr to register 0x34 on vcpu 1
```

**Comment by nlf** [comment]
it's not encoded, that's just what the cli returns. give it a minute and try again and it should work

**Comment by ruimarinho** [comment]
Looks like the docker service isn't starting inside the VM.

**Comment by nlf** [comment]
can you ssh into the VM? see if it's running

**Comment by ruimarinho** [comment]
yep, it isn't running - trying to debug the rc script

**Comment by nlf** [comment]
let me know what you find

**Comment by ruimarinho** [comment]
I can't resolve dns inside the vm. `wget: unable to resolve host address 'get.docker.io'`

I think it's pointing to my previous VM:

```
$ cat /etc/resolv.conf
192.168.64.1
nameserver 192.168.64.1 # eth0
```

**Comment by nlf** [comment]
what happens if you delete the top line?

**Comment by nlf** [comment]
you can also try specifying a dns server in the config, do

``` sh
dlite config
```

and set `dns_server` to `8.8.8.8`. it'll stop and restart the daemon for you. the missing "nameserver" at the beginning of the string is a bug though, working on that right now (though i'm curious what else you find is broken)

**Comment by ruimarinho** [comment]
Using `dlite config`:

```
8.8.8.8
nameserver 192.168.64.1 # eth0
```

Setting nameserver manually to 8.8.8.8 worked for me. What's running on 192.168.64.1?

**Comment by nlf** [comment]
that should be your host machine, which is why dhcp gives that for the eth0 device. interesting that it works for me (and a few other people i had test) but not for you.. can you ping 192.168.64.1 from inside the vm?

**Comment by ruimarinho** [comment]
Yep, it's reachable from inside the VM.

**Comment by nlf** [comment]
I'm guessing if you run `nslookup google.com 192.168.64.1` from inside your vm it fails?

**Comment by ruimarinho** [comment]
Yep, it fails. Could be something on my host machine, but I can correctly resolve dns on it.

**Comment by nlf** [comment]
that's really weird.. is your firewall turned on? do you have any adblockers that work on the dns level? or things like [hands off](https://www.oneperiodic.com/products/handsoff/) installed?

maybe i should put the default back to google's dns server

**Comment by ruimarinho** [comment]
Yay, it looks like my dnsmasq service had some interference with it. I've stopped it just to remove it from the equation and restarted the VM. It looks good now.

Found two more issues. When editing with `dlite config`, it often outputs this:

```
virtio_net: Could not create vmnet interface, permission denied or no entitlement?
```

Apparently it likes `sudo`.

The other issue is with extra args. I think it's not passing them to /proc/cmdline:

```
console=ttyS0 hostname=dlite uuid=7647d5a7-e32b-11e5-b178-c42c032ff040 dns_server=192.168.64.1 user_name=<username> user_id=501 docker_version=1.10.2 docker_extra=
```

**Comment by nlf** [comment]
it shouldn't need sudo, maybe it just needs a delay between stopping and starting or something. it runs the same function that `dlite start` does, which calls `launchctl stop local.dlite` without sudo.

that's strange about your extra args, it's definitely passing them to xhyve, maybe something else is dropping them. i'll poke around a bit

**Comment by nlf** [comment]
what did you try to pass as extra arguments?

**Comment by ruimarinho** [comment]
Yeah, I checked the code and it seems fine. There's something sketchy about stopping/starting dlite so it could just be the machine not picking up the new config because it's not getting restarted.

**Comment by nlf** [comment]
i did just find a small bug with extra parameters in the init scripts, i'll have a new test build of dhyve-os up in a few minutes

**Comment by ruimarinho** [comment]
Ah yeah, got it :) anything after a space will be ignored.

**Comment by ruimarinho** [comment]
What about keeping `cat /proc/cmdline | sed -n 's/^.*docker_extra=\(.*\)$/\1/p`?

**Comment by ruimarinho** [comment]
Found two more warnings - one on the host:

```
XHYVE: vlapic callout at 0x15948.0x597a03ab6634e9a1, expected at 0x15948.#59a953c7eb92360f
```

Another inside the VM:

```
time="2016-03-06T00:17:44.710000000Z" level=warning msg="Your kernel does not support cgroup blkio throttle.write_bps_device"
time="2016-03-06T00:17:44.710000000Z" level=warning msg="Your kernel does not support cgroup blkio throttle.read_iops_device"
time="2016-03-06T00:17:44.710000000Z" level=warning msg="Your kernel does not support cgroup blkio throttle.write_iops_device"
```

**Comment by nlf** [comment]
that's what i'm doing with the regex.

as for the warnings, the top one isn't anything to worry about. it's just a warning from xhyve which still outputs a fair amount of debugging information.

the warnings from docker are interesting though, i'll take a look now. could be i goofed something in the kernel config.

**Comment by ruimarinho** [comment]
The config command is having difficulties in restarting the agent. The `virtio_net` is always output to the log.

**Comment by nlf** [comment]
try manually stopping the agent and waiting a minute or so and then start it back up, maybe we just need to introduce a delay

**Comment by ruimarinho** [comment]
`dlite stop` is not working, although no output is shown. Should `/usr/local/bin/dlite` run as root?

**Comment by nlf** [comment]
not for the stop command, no. running `dlite stop` with sudo will definitely fail

**Comment by ruimarinho** [comment]
start and stop with patience (watching the logs in these case) works reliable (always without sudo). I think dlite config needs a delay or a more reliable way of finding out when the agent is stopped.

**Comment by nlf** [comment]
stop the service and run `dlite update -v v3.0.0` to pull the latest OS i just pushed, it resolves the nameserver issue, the docker extra args issue, and the cgroup warnings from the docker logs.

**Comment by nlf** [comment]
yeah i agree, i'll work something up that checks the output of `ps` to make sure the daemon is really stopped all the way.

**Comment by ruimarinho** [comment]
Looks good! Took me some tries to get the VM to update to the latest 3.0.0 commit.

Here's how resolve.conf appears:

```
$ cat /var/log/resolv.conf
nameserver 192.168.64.1
nameserver 192.168.64.1 # eth0
```

The extra args are parsing correctly now.

**Comment by nlf** [comment]
If you can build from master, try what I just pushed. I put in a loop with a delay to make sure the agent is stopped before StopAgent() returns, let's see if that fixes the config command for you.

**Comment by wbinnssmith** [comment]
hi @nlf! thanks so much for dlite and for 2.0 — super excited for virtfs.

after `dlite stop`, `dlite install 3.0.0`, `dlite update -v v3.0.0` and `dlite start`, `docker ps` seems to hang even after waiting for quite a bit for things to come up.

is sshing into the vm as straightforward as before? it now prompts for a password, and `docker` for neither the `docker` user nor `root` as on the dhyveos README seems to do the trick.

**Comment by nlf** [comment]
@wbinnssmith did you remove your old installation first?

the upgrade steps should be something like:
- run `dlite stop`
- run `sudo dlite uninstall`
- run `sudo nfsd stop`
- update your dlite binary
- run `sudo dlite install -v v3.0.0`
- run `dlite start`
- wait ~30 seconds

**Comment by wbinnssmith** [comment]
ah yep, didn't get the prerelease binary from releases. derp. check back in soon :)

**Comment by ruimarinho** [comment]
Building from master - I can reliable edit the config (stop works very well) but occasionally the `virtio_net` message appears and starting the machine takes longer (15 seconds vs 2-3 seconds). Any idea?

**Comment by nlf** [comment]
how occasionally? every other time? every third time? when it happens does the machine boot anyway or do you have to run `dlite start` again?

**Comment by ruimarinho** [comment]
I can't seem to find a pattern yet. Sometimes the machine takes longer to boot and eventually boots, but sometimes I have to `dlite start` it again. I'll see if I can replicate it more consistently.

**Comment by nlf** [comment]
also see if stopping it manually and waiting a while before starting it again helps

**Comment by wbinnssmith** [comment]
@nlf just wanted to let you know that after installing the prerelease binary, all is great. Love the addition to the ssh config, and p9 is working great with volume mounts :smile:

**Comment by nlf** [comment]
fantastic, thanks for letting me know @wbinnssmith

**Comment by ruimarinho** [comment]
Should `dlite uninstall` stop the daemon or isn't that really required? What about cleaning up `/var/db/dhcpd_leases`?

**Comment by ruimarinho** [comment]
`~/.ssh/config` is being written as root but should be kept owned by the current user.

**Comment by ruimarinho** [comment]
Is `/sbin/nfsd` still necessary on the sudoers file?

**Comment by ruimarinho** [comment]
Technically, `sudo dlite install` does not touch `~/.ssh/config`, only start does.

**Comment by ruimarinho** [comment]
Having some issues when pulling/extracting images via `docker-compose`:

```
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "compose/cli/main.py", line 56, in main
  File "compose/cli/docopt_command.py", line 23, in sys_dispatch
  File "compose/cli/docopt_command.py", line 26, in dispatch
  File "compose/cli/main.py", line 191, in perform_command
  File "compose/cli/main.py", line 524, in run
  File "compose/cli/main.py", line 711, in run_one_off_container
  File "compose/project.py", line 316, in up
  File "compose/service.py", line 352, in execute_convergence_plan
  File "compose/service.py", line 253, in create_container
  File "compose/service.py", line 280, in ensure_image_exists
  File "compose/service.py", line 771, in pull
  File "compose/progress_stream.py", line 18, in stream_output
  File "compose/utils.py", line 63, in split_buffer
  File "json/decoder.py", line 369, in decode
ValueError: Extra data: line 3 column 1 - line 238 column 1 (char 190 - 43311)
docker-compose returned -1
```

With `docker pull`, everything works fine (workaround for now).

**Comment by bpinto** [comment]
Hey, I am trying to get it to work. One thing I found weird is that as `docker` is now installed as part of a script, when running `docker ps` inside the VM, nothing is printed in case docker failed to download. Is there a way to come with docker pre-installed as before?

Edit: I guess to me it doesn't really matter if it comes pre-installed or not, just wondering what are the benefits of not having it pre-installed.

**Comment by bpinto** [comment]
Edit: Comment deleted.

**Comment by antoniocanas** [comment]
This release -as previous ones- mounts volumes with wrong permissions 501:dialout

**Comment by bpinto** [comment]
I'm also experiencing problems with permissions, but it's different from when we moved to `root`. The files permissions are set with my user permission `bpinto:staff` but the docker user is not being able to write to some files.

**Comment by nlf** [comment]
@antoniocanas can you tell me your user id and the group id of the 'staff' group from your osx host?

to get the group id, you should be able to run `dscl . read /Groups/staff` and see output similar to this

```
dscl . read /Groups/staff
AppleMetaNodeLocation: /Local/Default
GeneratedUID: ABCDEFAB-CDEF-ABCD-EFAB-CDEF00000014
GroupMembers: FFFFEEEE-DDDD-CCCC-BBBB-AAAA00000000
GroupMembership: root
Password: *
PrimaryGroupID: 20
RealName: Staff
RecordName: staff BUILTIN\Users
RecordType: dsRecTypeStandard:Groups
SMBSID: S-1-5-32-545
```

as for the user id, `echo $UID` should show that

**Comment by nlf** [comment]
@bpinto hmm.. i see the problem. let me see if i can come up with something to resolve that

**Comment by nlf** [comment]
@bpinto actually.. i haven't been able to replicate that, can you help me figure out how to replicate your issue so i can find a fix for it? i just tested files from my share that have `600` permissions meaning only my user (on the OSX host) has permissions to read or write it, and had no issues at all. what are you trying to do that's failing for you?

**Comment by bpinto** [comment]
@nlf I will try to create a reproducible scenario. I am running a rails server if you have rails installed and could run one server, it might have the same error.

**Comment by bpinto** [comment]
@nlf 

```
~ ◦ $ la log/
total 8
drwxr-xr-x   3 bpinto  staff   102B Mar  7 17:16 .
drwxr-xr-x  42 bpinto  staff   1.4K Mar  5 19:24 ..
-rw-r--r--   1 bpinto  staff     2B Mar  7 17:16 development.log

~ ◦ $ docker run -t -v $PWD/log:/log ruby:2.1.6 /bin/bash -lc "ls -lah log; echo a >> log/development.log"
total 1.0K
drwxr-xr-x 1  501 dialout 102 Mar  7 17:16 .
drwxr-xr-x 1 root root    180 Mar  7 17:16 ..
-rw-r--r-- 1  501 dialout   2 Mar  7 17:16 development.log
/bin/bash: log/development.log: Operation not permitted
```

Oh `501:dialout`... I hadn't noticed from inside the container the files had different permissions... Weird...

**Comment by nlf** [comment]
`501:dialout` is because while i map the user id between osx and the vm, nothing maps the user between the vm and the containers

**Comment by nlf** [comment]
ok, i'm able to reproduce it now. i'll see if i can come up with a fix

**Comment by nlf** [comment]
ok. so, we're now back to a point where write permissions within the container are a non issue. new files created in mounted volumes are owned by root in the host operating system, however. i'm still investigating to see if there's a better way to resolve that.

the docker binary is back to having the latest version pre-installed which makes for a faster initial boot unless you're requesting something other than the latest version of docker (or if a new docker release comes out and the image hasn't been rebuilt yet, i'm looking in to setting up a hook so the image will be kept up to date as well as possible)

also increased the size of the socket buffers, which should improve speed for things like `docker logs`

**Comment by antoniocanas** [comment]
@nlf 
It still shows `drwxr-xr-x   1  501 dialout   170 Mar  8 15:46 app`

The official mysql image, for example, will not work:

```
db_1 | 2016-03-09T11:12:19.857765Z 0 [ERROR] InnoDB: Operating system error number 1 in a file operation.
db_1 | 2016-03-09T11:12:19.857883Z 0 [ERROR] InnoDB: Error number 1 means 'Operation not permitted'
db_1 | 2016-03-09T11:12:19.857909Z 0 [ERROR] InnoDB: File ./ibdata1: 'stat' returned OS error 101.
db_1 | 2016-03-09T11:12:19.857922Z 0 [ERROR] InnoDB: os_file_get_status() failed on './ibdata1'. Can't determine file permissions
```

**Comment by danquah** [comment]
Yup, I'm getting the same error with the official mariadb image, it does seem its hitting a slightly different problem:

Directly from osx:

```
$ ls -ln /Users
total 0
drwxr-xr-x+  11 201 201  374 Nov  9 07:53 Guest
drwxrwxrwt    7   0   0  238 Nov  9 13:25 Shared
drwxr-xr-x+ 101 501  20 3434 Mar  9 15:25 danquah
```

From inside dhyveos as the docker user

```
$ ls -ln /Users
ls: /Users/danquah: Operation not permitted
total 0
```

And as root

```
$ sudo ls -ln /Users
total 4
drwxr-xr-x    1 501      20            3434 Mar  9 14:25 danquah
```

How does the permissions around the mountpoint work? Seems the docker user (uid 1000) is not allowed to even list the mounted volumne?

**Comment by nlf** [comment]
this file permissions thing is a real pain. i'm still trying to figure out the best way to resolve this. please do keep reporting specific use cases that fail, it gives me more things to test against while i try to find a solution.

**Comment by danquah** [comment]
@nlf just out of curiosity, what problem did the access=any + uname settings in https://github.com/nlf/dhyve-os/commit/0b1f7233e68d7e0cd5671c36b6e6801c05cabe19#diff-37aa5e55ffbda42c8c4d48f5feb1be5fL50 cause? I just remounted with those options and I'm getting a bit further. Files created from within dhyveos is now created as my user, but files created by a running container (even deeper within dhyveos I guess) are now appearing with uids from the container (in this case uid=999) which cases new kinds of problems.

**Comment by nlf** [comment]
it fixed some issues where permissions were being denied. i still haven't found a way to make everything happy.

**Comment by danquah** [comment]
hm, might be the problems I'm seeing now: while the mariadb container is able to create its mysql-directory with uid=999 any attempts at doing anything inside that directory fails. It seems that the uid=999 comes from explicit chowns done inside the container to make sure the mysql data-directory is owned by the mysql-user - and those ids make it all the way back to macos.

Files created without trying to force a uid/gid seems to be created as as my local mac-user - so that part at least works.

Any idea where the mapping between the hostos' file-system uids/guids and the fs presented to the container happens?

**Comment by nlf** [comment]
right now there's no mapping. i'm still trying to figure out if there's a safe way of forcibly mapping everyone to the host's user

**Comment by ruimarinho** [comment]
What about `--userns-remap`? Would that be of any help here? E.g. https://github.com/boot2docker/boot2docker/pull/1142

**Comment by danquah** [comment]
As I understand docker and user namespaces it won't help here. Namespacing allows the docker deamon to map uids/gids used inside a container to another set of id's outside the container. eg, for container 1 uid 0 is mapped to 10000 in the hostos (uid 1000 to 11000 and so forth), and for container 2 uid 0 is mapped to 20000, uid 1000 to 21000.

What would be nice in the dlite-case (as I understand it) would be to squash everything to the macos user running dlite. The problem then is the world view being presented to the container. It would probably see uid/gids from the hostos instead of its own, and that might cause problems if it tries to enforce a specific uid/gid.

**Comment by nlf** [comment]
correct, user namespaces won't resolve the issue here

**Comment by nlf** [comment]
and double correct, squashing all permissions to the host uid/gid is problematic

**Comment by antoniocanas** [comment]
So, right now, there's no way to use Docker on OSX, is it? dinghy, docker-machine-nfs or any other solution have the same permissions problem.

**Comment by nlf** [comment]
i just spent a bunch of time hacking up the 9p library to allow for different ownership on the host than there is in the guest, and it seems to be working.

by that, i mean i was able to run the mysql container pointing to a directory on my host. files were all created on the host as my user. files in the container were owned by `mysql:staff` which seemed to be good enough to make mysql happy.

i'll have a new test build up soon

**Comment by danquah** [comment]
Awesome! Be happy to test it of cause :)

**Comment by nlf** [comment]
ok folks, beta4 of 2.0.0 is up and should resolve at least some of the permissions issues between host and guest.

what i did was hack the 9p filesystem support in xhyve to do a few things differently:
- allow connections from anyone (this means that the docker user can `ls` in `/Users/youruser` now, though still won't be able to read/write files unless they're globally accessible). this also means that if a container does, for example `chown mysql:mysql /var/lib/mysql` later reads will actually be performed as the `mysql` user instead of as the host's user, this is what allowed mysql to start working correctly
- instead of changing ownership of a file on the host when a `chown` call is made on the guest, the new ownership information is written to the file's extended attributes on the host
- as a partner for the above change, `stat` calls inside the guest will be given the ownership information stored in the extended attributes instead of the actual ownership. if no extended attributes exist, the host's ownership will be relayed
- all files are written to the host as the user running dlite no matter what

this release also takes a shot at resolving the weird buffering issues for some commands, like `logs -f`

as a heads up, in my above example where `chown mysql:mysql /var/lib/mysql` is done, the ownership in the guest will appear as `mysql:staff`. this is because 9p doesn't request a GID when the filesystem is attached so i have no way of determining the correct group to map to, which means i fall back to the primary group of the host's user. so far from my limited testing this is only cosmetic as the majority of services don't particularly care about group permissions.

as always, please post feedback you have whether it's positive or negative. it's all helpful :)

**Comment by antoniocanas** [comment]
Thanks for your effort :)

Testing it, docker is not starting:

```
~/tmp
❯ sudo ./dlite uninstall
Removing launchd agent: done
Removing files: done

~/tmp
❯ sudo ./dlite install -v 3.0.0-beta5
...
Would you like to continue? (Y/n):
Building disk image: done
Downloading OS: done
Generating SSH key: done
Writing configuration: done
Creating launchd agent: done

~/tmp 40s
❯ ./dlite start
Starting the agent: done
The VM may take some additional time to fully boot

~/tmp
❯ date
Mon Mar 14 21:52:20 CET 2016

~/tmp
❯ docker ps
Cannot connect to the Docker daemon. Is the docker daemon running on this host?

~/tmp
❯ date
Mon Mar 14 21:57:08 CET 2016

~/tmp
❯ docker ps
Cannot connect to the Docker daemon. Is the docker daemon running on this host?
```

**Comment by nlf** [comment]
yeah dhyve-os beta5 was a bust, sorry about that. you should be able to use beta4 and i already took down beta5

**Comment by antoniocanas** [comment]
Same problem with beta4

**Comment by nlf** [comment]
you did a full uninstall reinstall cycle? let me see if i can replicate here

**Comment by nlf** [comment]
i just did the following:

```
dlite stop
sudo dlite uninstall
sudo dlite install -v v3.0.0-beta4
dlite start
# wait ~30 seconds
docker ps
```

with success, maybe you're having another issue. are you able to ssh to the vm with `dlite ssh`?

**Comment by antoniocanas** [comment]
Edit: reinstalled (after reboot) again and now it's working :)

**Comment by nlf** [comment]
are you able to ssh to the vm? `dlite ssh` should get you in

**Comment by antoniocanas** [comment]
Yeah it's working now, sorry.
Again, same permissions problem:
from osx host: `drwxr-xr-x  23 antonio  staff   782B Mar 14 22:31 app`
from container: `drwxr-xr-x   1  501 dialout  782 Mar 14 21:31 app`

**Comment by danquah** [comment]
Think I had the same problem, I fixed it by doing a dlite config and changing the dns_server to 8.8.8.8.
For some reason dhyveos could not do any lookups, and it resulted in a 0-byte /usr/bin/docker

Would it be possible to run some kind of diagnostics (eg "dlite verify") after the deamon is up and running that could check things like this?

I have docker up and running now, and the mariadb container I'm figthing with is now creating files and directories owned by my mac user, and from inside the container the same files shows up owned by mysql:mysql.

It still runs in to some permission problems though, I'll debug som more.

**Comment by nlf** [comment]
it will _always_ show `501 dialout` as the permissions for files in your user directory, because that's who they belong to. there's no user with id `501` in the container, so that gets displayed numerically, however there is a group with id `20` named `dialout` so you see that one as a string.

the changes i made allows things like `chown` to work correctly as far as the vm is concerned, for example in your container `chown root:root` a file at random and you'll see in the container it shows as owned by root (though the group will still be dialout), however on your host it will still be owned by your user

**Comment by nlf** [comment]
@danquah yeah a `verify` or `troubleshoot` command or something is a good idea

what permission problems are you seeing it run into? i was able to run a mariadb container though admittedly i never actually created a table or wrote any data to it.

**Comment by antoniocanas** [comment]
So do I need to inherit each image Dockerfile's and 'chown' the volume?

**Comment by nlf** [comment]
no, you just need to use the files. the display of `501 dialout` is only cosmetic. your containers should all work fine.

**Comment by antoniocanas** [comment]
I see, the UID is correct. Thanks a lot :)

**Comment by danquah** [comment]
@nlf it gets started, populates its datadir with some basic files, and then fails.

To reproduce:

```
mkdir ~/dbtest
docker run -i --name mariadb-test -v ~/dbtest:/var/lib/mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mariadb:5
docker logs mariadb-test 
```

Where I amongst other messages found

```
Initializing database
160314 22:39:14 [Note] /usr/sbin/mysqld (mysqld 5.5.48-MariaDB-1~wheezy) starting as process 59 ...
ERROR: 1005  Can't create table 'db' (errno: 13)
160314 22:39:14 [ERROR] Aborting
```

Permissions from a container:

```
$ docker run -it -v ~/dbtest:/var/lib/mysql mariadb:5 ls -l /var/lib/mysql
total 17
-rw-r----- 1 mysql dialout 16384 Mar 14 22:39 aria_log.00000001
-rw-r----- 1 mysql dialout    52 Mar 14 22:39 aria_log_control
drwx------ 1 root  dialout    68 Mar 14 22:39 mysql
```

And from osx

```
$ ls -l ~/dbtest
total 20
-rw-r----- 1 danquah staff 16384 Mar 14 23:39 aria_log.00000001
-rw-r----- 1 danquah staff    52 Mar 14 23:39 aria_log_control
drwx------ 2 danquah staff    68 Mar 14 23:39 mysql
```

Don't know yet whether that mysql-directory is supposed to be owned by root, might be the problem ....

**Comment by nlf** [comment]
yeah i kind of suspect that's the case.. i'm not sure why mariadb wouldn't chown the directory itself though. i'll take a look.

**Comment by danquah** [comment]
Ok, this is wird, it seems that I can't fix the ownership manually:

```
docker run -i --rm -v ~/dbtest:/var/lib/mysql mariadb:5 /bin/bash -c "ls -l /var/lib/mysql; chown -v mysql /var/lib/mysql/mysql; ls -l /var/lib/mysql"

total 17
-rw-r----- 1 mysql dialout 16384 Mar 15 08:40 aria_log.00000001
-rw-r----- 1 mysql dialout    52 Mar 15 08:40 aria_log_control
drwx------ 1 root  dialout    68 Mar 15 08:40 mysql
changed ownership of `/var/lib/mysql/mysql' from root to mysql
total 17
-rw-r----- 1 mysql dialout 16384 Mar 15 08:40 aria_log.00000001
-rw-r----- 1 mysql dialout    52 Mar 15 08:40 aria_log_control
drwx------ 1 root  dialout    68 Mar 15 08:40 mysql
```

Quick script to replicate:

```
#!/usr/bin/env bash -x
docker rm mariadb-5-init

rm -fr ~/dbtest
mkdir ~/dbtest
docker run -i --name mariadb-5-init -v ~/dbtest:/var/lib/mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw mariadb:5
docker run -i --rm -v ~/dbtest:/var/lib/mysql mariadb:5 /bin/bash -c "ls -l /var/lib/mysql; chown -v mysql /var/lib/mysql/mysql; ls -l /var/lib/mysql"
```

**Comment by antoniocanas** [comment]
Didn't look inside the container, but the official mysql is working out of the box for me

**Comment by danquah** [comment]
@antoniocanas - good point, I just did the same test with mysql:latest and it does indeed work.
Directory-listing after the container came up

```
-rw-r----- 1 mysql mysql         56 Mar 15 09:20 auto.cnf
-rw-r----- 1 mysql mysql       1319 Mar 15 09:21 ib_buffer_pool
-rw-r----- 1 mysql mysql   50331648 Mar 15 09:21 ib_logfile0
-rw-r----- 1 mysql mysql   50331648 Mar 15 09:20 ib_logfile1
-rw-r----- 1 mysql mysql   79691776 Mar 15 09:21 ibdata1
-rw-r----- 1 mysql dialout 12582912 Mar 15 09:21 ibtmp1
drwxr-x--- 1 mysql mysql       2618 Mar 15 09:20 mysql
drwxr-x--- 1 mysql mysql       3060 Mar 15 09:20 performance_schema
drwxr-x--- 1 mysql mysql       3672 Mar 15 09:20 sys
```

The problem seems to be that once a file has a ownership, it cannot be changed:

```
root@5a9a1c0d470f:/var/lib/mysql# touch /var/lib/test_outside_volume

root@5a9a1c0d470f:/var/lib/mysql# ls -l /var/lib/test_outside_volume
-rw-r--r-- 1 root root 0 Mar 15 09:26 /var/lib/test_outside_volume

root@5a9a1c0d470f:/var/lib/mysql# chown -v mysql /var/lib/test_outside_volume
changed ownership of '/var/lib/test_outside_volume' from root to mysql

root@5a9a1c0d470f:/var/lib/mysql# ls -l /var/lib/test_outside_volume
-rw-r--r-- 1 mysql root 0 Mar 15 09:26 /var/lib/test_outside_volume

root@5a9a1c0d470f:/var/lib/mysql# touch /var/lib/mysql/test_inside_volume

root@5a9a1c0d470f:/var/lib/mysql# ls -l /var/lib/mysql/test_inside_volume
-rw-r--r-- 1 root dialout 0 Mar 15 09:31 /var/lib/mysql/test_inside_volume

root@5a9a1c0d470f:/var/lib/mysql# chown -v mysql /var/lib/mysql/test_inside_volume
changed ownership of '/var/lib/mysql/test_inside_volume' from root to mysql

root@5a9a1c0d470f:/var/lib/mysql# ls -l /var/lib/mysql/test_inside_volume
-rw-r--r-- 1 root dialout 0 Mar 15 09:31 /var/lib/mysql/test_inside_volume
```

Did the same test for at file owned by mysql trying to change the ownership back to root, same result.

**Comment by antoniocanas** [comment]
Disk size argument it's not working:

```
~/tmp
❯ sudo ./dlite install -v v3.0.0-beta4 --disk=100
Password:
The install command will make the following changes to your system:
- Create a '.dlite' directory in your home
- Create a 100 GiB sparse disk image in the '.dlite' directory
...

~/tmp 16s
❯ ./dlite start
Starting the agent: done
The VM may take some additional time to fully boot

~/tmp
❯ ssh docker@local.docker
Warning: Permanently added '192.168.64.9' (ECDSA) to the list of known hosts.
DhyveOS version 3.0.0
Docker version 1.10.3, build 20f81dd
$ df -h
Filesystem                Size      Used Available Use% Mounted on
devtmpfs                995.2M         0    995.2M   0% /dev
tmpfs                  1001.9M         0   1001.9M   0% /dev/shm
tmpfs                  1001.9M     40.0K   1001.9M   0% /tmp
tmpfs                  1001.9M     36.0K   1001.9M   0% /run
/dev/sda1                49.1M     34.2M     11.1M  75% /mnt/overlay
overlay                  49.1M     34.2M     11.1M  75% /etc
overlay                  49.1M     34.2M     11.1M  75% /usr/bin
/dev/sda3                19.0G   1003.7M     17.8G   5% /var/lib/docker
none                   1001.9M         0   1001.9M   0% /sys/fs/cgroup
/dev/sda3                19.0G   1003.7M     17.8G   5% /var/lib/docker/btrfs
```

I've tried to create a bigger disk because PHP's Composer throws this
`The disk hosting /app/vendor is full, this may be the cause of the following exception`

I guess space is not the problem, because I can dd a large zero-filled file, probably PHP is getting the amount of free space and fail for a sparsed disk.

It's a volume shared with the host, my docker-compose.yml:

```
app:
  image: tianon/true
  volumes:
    - ./app:/app
```

**Comment by nlf** [comment]
@antoniocanas hmm.. i'm unable to reproduce this one. does it work for you if you do:

``` sh
dlite stop
dlite rebuild -d 100
dlite start
```

**Comment by nlf** [comment]
@danquah checking it out to see if i can reproduce it on my end, thanks for the detailed example

**Comment by nlf** [comment]
@danquah so here's a weird thing.. `chown mysql /var/lib/mysql/mysql` fails, but `chown mysql:mysql /var/lib/mysql/mysql` works fine. i'm adding some debugging on my end so i can try to figure out what mariadb is doing to see if i can fix it

**Comment by antoniocanas** [comment]
@nlf I also did that and same problem.
I'll try at home later :)

**Comment by nlf** [comment]
when you get home, after you run `dlite stop` but before you try to rebuild, check the output of `hdiutil info` and see if there's an entry for `disk.sparseimage` in there. i know the rebuild doesn't work correctly if the old disk didn't get unmounted somehow, so possibly there's a bug there

**Comment by nlf** [comment]
@danquah fixed :) check out 2.0.0-beta5

**Comment by antoniocanas** [comment]
@nlf Yes there is

```
./dlite stop
Stopping the agent: done

~/tmp
❯ hdiutil info
framework       : 417.1
driver          : 10.11v417.1
================================================
image-path      : /Users/antonio/.dlite/disk.sparseimage
image-alias     : /Users/antonio/.dlite/disk.sparseimage
shadow-path     : <none>
icon-path       : /System/Library/PrivateFrameworks/DiskImages.framework/Resources/CDiskImage.icns
image-type      : sparse disk image
system-image    : false
blockcount      : 41943040
blocksize       : 512
writeable       : TRUE
autodiskmount   : false
removable       : TRUE
image-encrypted : false
mounting user   : root
mounting mode   : <unknown>
process ID      : 491
/dev/disk2  FDisk_partition_scheme
/dev/disk2s1    Linux
/dev/disk2s2    Linux_Swap
/dev/disk2s3    Linux
```

**Comment by nlf** [comment]
@antoniocanas is it still there after a `dlite rebuild -d 100`?

**Comment by antoniocanas** [comment]
Yes

**Comment by nlf** [comment]
ah ha :) for some reason detaching the volume isn't working correctly, what do you get if you run `hdiutil detach /dev/disk2`?

**Comment by antoniocanas** [comment]
```
❯ hdiutil detach /dev/disk2
"disk2" unmounted.
hdiutil: couldn't eject "disk2" - Resource busy
```

**Comment by nlf** [comment]
and you're sure dlite has stopped? check `ps aux | grep dlite`

**Comment by antoniocanas** [comment]
This is strange...

```
❯ ./dlite stop
Stopping the agent: ERROR!
The agent is already stopped

~/tmp
❯ ps aux | grep dlite
root              483   3.1  0.3 575542528  28784   ??  S     6:43PM   3:07.08 /Users/antonio/tmp/dlite daemon
antonio          4699   0.0  0.0  2423976    292 s001  R+    8:16PM   0:00.00 grep --color=auto dlite
```

**Comment by nlf** [comment]
ahh, it didn't die for some reason.. you can do `sudo kill 483` and wait a minute to see if it dies, if it doesn't do `sudo kill -9 483` to kill it forcibly

**Comment by antoniocanas** [comment]
Right, now it's working, Thanks a lot

**Comment by danquah** [comment]
:clap: works perfectly, I can now boot my mariadb container. I'll test some more on my day-to-day setup and see if I can get into any more trouble :)

**Comment by mikz** [comment]
I have an issue with the dhyve-os image (v3.0.0-beta4).

Docker does not start at all, so when i ssh to the machine I can see it is not running. When i try to run the init.d it says ok, but does not start. After some digging I found out that docker binary has size 0:

```
$ ls -lA /usr/bin/docker
-rwxr-xr-x    1 root     root             0 Mar 15 18:59 /usr/bin/docker
```

I tried to fix that by downloading a docker release, but the internet does not work inside the machine.

```
$ ping google.com
ping: bad address 'google.com'
$ cat /etc/resolv.conf
nameserver 192.168.64.1
nameserver 192.168.64.1 # eth0
```

Where 192.168.64.1 is my machine. But I guess it does not run dns server?

```
dig google.com @192.168.64.1                              ruby-2.3.0

; <<>> DiG 9.8.3-P1 <<>> google.com @192.168.64.1
;; global options: +cmd
;; connection timed out; no servers could be reached
```

Fixed that by reinstalling dlite with `--dns-server=8.8.8.8`.
But then dlite does not really start. On first attempt I can't ssh to it and on second it keeps starting the agent. And I could find in the logs:

> virtio_net: Could not create vmnet interface, permission denied or no entitlement?

**Comment by mikz** [comment]
Looks like I forgot to reinstall it with `-v 3.0.0-beta4`. After doing that and using `--dns-server=8.8.8.8` it works correctly!

**Comment by nlf** [comment]
@mikz good to know! i'm still not sure why dns doesn't work correctly out of the box for some people, that's why the --dns-server option exists :) i should probably document that better though

**Comment by mikz** [comment]
AFAIK OSX will start DNS server if you have Internet Sharing enabled in System Preferences > Sharing.

**Comment by nlf** [comment]
I don't have that enabled and dns works fine for me. I do know that some people have had dnsmasq running on their host system and that has caused issues. Still haven't been able to see if there are other situations that cause a failure there.

**Comment by antoniocanas** [comment]
I still have problems with PHP's Composer:

```
  - Installing symfony/console (v3.0.3)
    Downloading: 100%
    Failed to download symfony/console from dist: Could not delete /app/lumen/vendor/composer/254938c2:
    Now trying to download from source
  - Installing symfony/console (v3.0.3)
    Cloning 2ed5e2706ce92313d120b8fe50d1063bcfd12e04
The disk hosting /app/lumen/vendor is full, this may be the cause of the following exception
```

**Comment by mikz** [comment]
Thats it! I have dnsmasq running on `::1:53` and `127.0.0.1:53`.  I don't really want to run it on all the interfaces. 

Guess some dns proxy running on that interface and random port would make sense. Like http://pow.cx/docs/dns_server.html. But initially dlite could check if there is dns server running and print a warning or something.

**Comment by nlf** [comment]
@antoniocanas can you verify with `df -h` from an ssh session to the vm that there is disk space? if there is, can you reproduce this in a small test scenario that you can send me so i can try to figure out what's going on?

**Comment by nlf** [comment]
@mikz excellent! i'll have to see what i can do. if i can detect a process listening on port 53 i can print a warning that should help other users not have to deal with this

**Comment by antoniocanas** [comment]
@nlf There is enough space. You could test it by just running:
`docker run --rm -v $(pwd):/app composer/composer require symfony/console`

**Comment by nlf** [comment]
that doesn't work for me, i get an error when it tries to clone `https://github.com/symfony/console.git`

**Comment by nlf** [comment]
oh. nevermind, i see. that's the error you're getting too :) i just noticed the disk full error at the top

**Comment by antoniocanas** [comment]
Yep. It works if you remove the -v argument, so it's someway related to the volume

**Comment by nlf** [comment]
it also works if you run the container with bash as the entrypoint and clone the repos in your home directory, then copy them to the /app/vendor directory. it appears something about composer doesn't play nicely with 9pfs

**Comment by nlf** [comment]
ok, looks like i may not be correctly reporting space. i'm going to look into the 9p spec and see if i can figure out what i need to change to support this.

**Comment by antoniocanas** [comment]
@nlf It's weird that just some kind of php packages fails, also composer has a 'diagnose' argument that shows: `Checking disk free space: OK`

Maybe the clue is in this line of the error:
`Failed to download symfony/console from dist: Could not delete /app/lumen/vendor/composer/254938c2`

**Comment by nlf** [comment]
it's all a side effect of the 9p server not reporting the free disk space correctly. i'm reading through protocol docs to try to find a fix

**Comment by STRML** [comment]
Anyone seeing their containers self-destruct on reboot/hard shutdown with 2.0.0? My Macbook has a "cold bug" (shuts down below an internal temp of roughly 20 degrees Celsius (yes, really)) and dlite appears to just rebuild when the system comes back up. Not a big deal since my containers are small, but could be a bigger pain in the future.

**Comment by nlf** [comment]
@STRML i have not seen that.. i do know that a hard shut down can cause corruption on the vm's disk (just like it can on your host's disk, ouch) but i've definitely never seen it just clear out everything

**Comment by STRML** [comment]
Yeah, daily kernel panics are not fun - out of warranty. I'll keep an eye on it, sounds like it's not something dlite could reasonably handle. Thanks.

**Comment by nlf** [comment]
just so everyone is aware, the free disk space issue seems to be inherent with 9p2000 protocol shares as there's no "statfs" message sent to the server, which is what would inform the guest of free disk space. there's a 9p2010 spec that adds this, but it looks like support in the kernel is lacking so i'm at a bit of an impasse.

my current plan is to evaluate embedding a usermode nfs server into dlite, with some changes to allow for better user mapping from container/vm to host (basically just using extended attributes like my 9p changes did). it may take a while and i'm not even sure if it'll work, but it's worth taking a look to see if it's even feasible.

**Comment by antoniocanas** [comment]
with official node:

```
npm ERR! Linux 4.4.3-dhyve
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "--save" "hapi"
npm ERR! node v5.4.0
npm ERR! npm  v3.3.12
npm ERR! path /app/node_modules/.staging/hoek-4b4bf751ce5c2289f71352c6623520d9
npm ERR! code EXDEV
npm ERR! errno -18
npm ERR! syscall rename

npm ERR! EXDEV: cross-device link not permitted, rename '/app/node_modules/.staging/hoek-4b4bf751ce5c2289f71352c6623520d9' -> '/app/node_modules/hapi/node_modules/hoek'
```

**Comment by nlf** [comment]
yeah the 9p share has all kinds of issues unfortunately. it's gonna be going away before 2.0.0 is officially released

**Comment by michaelshobbs** [comment]
For those running `2.0.0-beta5` with the firewall on, you will need to allow connections to bootp in order for the vm to startup

**Comment by danquah** [comment]
About that - did the 2.0.0-beta5 tag on https://github.com/nlf/dhyve-os/releases disappear a while back? Just attempted to do a re-install and it failed hard when it could not find the release :/

**Comment by nlf** [comment]
if you mean 3.0.0-beta5, then yes. i removed it because it had a severe bug. 3.0.0-beta4 is what you should be using

**Comment by STRML** [comment]
What was the severe bug?

**Comment by nlf** [comment]
to be completely honest, i don't remember. i do recall that i pulled the release within an hour or so of pushing it, though

**Comment by danquah** [comment]
ah yes, 3.0.0-beta5 - no problem then :)

Been running on 3.0.0-beta4 for a while now, it's pretty stable. My main issue currently is filesystem performance. I'm mainly using Docker for Drupal development, and Drupal has a habit of doing _a lot_ of file-stat operations. As they seem to be slow even running through 9p, the bulk of a the execution of a page-request ends up being php checking whether files exists.

Looking forward to seeing if you come up with something clever for (dlite) 2.0, and/or whether this mythical docker for Mac I'm signed up for (and almost nobody seems to have had access to?) has cracked the problem.

**Comment by nlf** [comment]
i'm working on implementing something similar to 9p but purpose built for dlite to ensure correct permissions. i have something that functions right now and am working on increasing performance. any examples you can provide of things that perform poorly with 9p would be appreciated, so i can test them against what i'm working on

**Comment by danquah** [comment]
Sure thing, I'll see if I can whip something quick (slow that is) up

**Comment by cmosetick** [comment]
@danquah I have some ideas for improving file system performance, which I think will help. I've been stuck in proof of concept mode though, as I've been really short on time the last month. No ETA right now, but I'll try to make some time.

**Comment by danquah** [comment]
I've done a simple test that has been packaged into an image that should be quite easy to run. 
The test just creates a lot of 1KB files runs a stat on them and reads them back in you should be able to run a test that uses a data-volume like this

```
docker run --rm -v ~/storage:/storage danquah/php-fileperf
```

You can find a simple script for running the test [in the repo](https://raw.githubusercontent.com/danquah/docker-php-fileperf/master/run.sh)

The test took about twice as long to run under dlite2:

Dlite1:

```
Creating 10000 files 1KB each in /storage, total size 10,000KB 
Created 10000 files in 11.88 ms
Got status of 10000 files in 1.08 ms 
Read 10000 files in 2.49 ms 
** Total duration: 15.45 **
```

Dlite 2:

```
Creating 10000 files 1KB each in /storage, total size 10,000KB 
Created 10000 files in 16.21 ms
Got status of 10000 files in 4.25 ms 
Read 10000 files in 17.4 ms 
** Total duration: 37.86 **
```

And just as a baseline - if I let the file-creation happen inside the containers without using volumes we have no problem:
Dlite 1

```
Creating 10000 files 1KB each in /storage, total size 10,000KB 
Created 10000 files in 0.5 ms
Got status of 10000 files in 0.04 ms 
Read 10000 files in 0.16 ms 
** Total duration: 0.7 **
```

Dlite 2

```
Creating 10000 files 1KB each in /storage, total size 10,000KB 
Created 10000 files in 0.76 ms
Got status of 10000 files in 0.04 ms 
Read 10000 files in 0.1 ms 
** Total duration: 0.9 **
```

I'll also try to do a test with a drupal-install - but in my first attempt I ran into permission-issues with dlite1 (drupal fails a deleting some files during install) - I'll be back :)

**Comment by danquah** [comment]
Got the Drupal install-test working. Runscript: https://raw.githubusercontent.com/danquah/docker-drupal-perf-test/master/run.sh

Results:

Dlite1, without volume

```
real    0m53.420s
user    0m4.200s
sys 0m1.330s
```

With volume

```
real    0m41.150s
user    0m5.220s
sys 0m33.560s
```

In other words - using a volume actually speeded up the install - weird.

Dlite 2
without volume

```
real    0m6.634s
user    0m3.900s
sys 0m0.660s
```

With a volume

```
real    0m16.701s
user    0m4.610s
sys 0m0.880s
```

So all of the sudden dlite2 is faster - even weirder when taking the previous test into consideration but obviously the installation is not hitting the same bottleneck as the basic file-creation test.

**Comment by rhesus** [comment]
I have followed the instructions in this issue and I am getting a different error when trying out master of dlite, when I ssh into the vm I see this:

`DhyveOS version 3.0.0`
`-sh: /etc/profile.d/dhyve.sh: docker: Text file busy`

Not sure what to do about it, the docker command is completely unresponsive and the logs don't really show anything interesting.
`2016/04/11 14:37:50 http: response.WriteHeader on hijacked connection`
`operation not supported by device`
`rdmsr to register 0x34 on vcpu 1`

**Comment by nlf** [comment]
that message is because docker hasn't finished downloading in the vm. give it a minute and check again

**Comment by rhesus** [comment]
Yes that was it, I wasn't patient enough.  Thanks for all of your hard work on dlite!!!

**Comment by nlf** [comment]
this thread is getting suuuuuuper long so i'm going to close it. let's just open new issues to discuss bugs, and i'll open new issues when i add new features as well
